docs/source/training.rst

```diff
@@ -0,0 +1,276 @@
+
+.. code-block:: python
+
+    from transformers import BertTokenizer, BertForSequenceClassification, AdamW
+
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+    model = BertForSequenceClassification.from_pretrained('bert-base-uncased')
+    model.train()
+    optimizer = AdamW(model.parameters(), lr=5e-5) # any PT optimizer works
+
+    text_batch = ["I love Pixar.", "I don't care for Pixar."]
+    labels = torch.tensor([1,0]).unsqueeze(0)
+    encoding = tokenizer.batch_encode_plus(text_batch, return_tensors='pt',
+                                           pad_to_max_length=True)
+    input_ids = encoding['input_ids']
+    attention_mask = encoding['attention_mask']
+
+    outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
+    loss = outputs[0]
+    loss.backward()
+    optimizer.step()
+
+Of course, you can train on GPU by calling ``to('cuda')`` on the model and
+inputs as usual.
+
+We also provide a few tools to make it easier to apply more complex
+fine-tuning procedures. Here's an example using weight decay and learning
+rate scheduling (omitting the data preparation steps for brevity):
+
+.. code-block:: python
+
+    from transformers import get_linear_schedule_with_warmup
+
+    device = torch.device('cuda')
+    model = BertForSequenceClassification.from_pretrained('bert-base-uncased').to(device)
+    model.train()
+
+    # Prepare optimizer and schedule (linear warmup and decay)
+    no_decay = ['bias', 'LayerNorm.weight']
+    optimizer_grouped_parameters = [
+        {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
+        {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
+    ]
+    optimizer = AdamW(optimizer_grouped_parameters, lr=1e-5)
+    scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_train_steps)
+
+    # iterate over batches of preprocessed training data
+    for batch in train_loader:
+        optimizer.zero_grad()
+        
+        input_ids = batch.input_ids.to(device)
+        attention_mask = batch.attention_mask.to(device)
+        labels = batch.labels.to(device)
+
+        outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
+        loss = outputs[0]
+        loss.backward()
+        optimizer.step()
+        scheduler.step()
+
+We highly recommend using :func:`~transformers.Trainer`, discussed below, which conveniently
+handles the moving parts of training Transformers models with features like
+mixed precision and easy tensorboard logging.
+
+
+Freezing the encoder
+--------------------
+
+In some cases, you might be interested in keeping the weights of the
+pre-trained encoder frozen and optimizing only the weights of the head
+layers. To do so, simply set the ``requires_grad`` attribute to ``False`` on
+the encoder parameters, which can be accessed with the ``base_model``
```

Comment:
I would add a remark that `base_model` works for every task-specific model in the library. It's a useful convention for the reader to know.